### PR TITLE
Adds chance of gas-spewing fissure when pickaxing basalt floor

### PR DESCRIPTION
**Pull request by read-0nly**
## Changelog
:cl:
add: Adds chance that using pickaxe on asteroid turf opens fissure
add: Fissures spew out various gas mixes
/:cl:

## About The Pull Request

Sometimes when a mommy pickaxe and a daddy basalt ok you know what I've had to do this joke three times now because I can't figure out git so screw it.

Use pickaxe on ground, sometimes ground cracks, hello gas maybe hello fire. Longer version is I modified Dig to on random chance change the basalt to a new icon fissure, who's dig icon is a badly made pickaxe hole and who's initial gasmix is one of many mixes picked at random with boring old hot co2 being most common for balance reasons but also not great if another one opens nearby with plasma or tritium.

Probably needs some massaging of the numbers and gas mixes to be "balanced"

## Why It's Good For The Game

I'm hoping it'll at least be funny
Makes rarer gasses a bit more accessible for fun
Makes science and engineering love mining more
Occasionally deepfries up miners

### CODE REVIEW

**Comment by read-0nly** [comment]
Hey I did it! I made the pull request!

File: code/game/turfs/simulated/floor/plating/asteroid.dm [hunk 1]
```diff
@@ -68,7 +68,27 @@
 			if(W.use_tool(src, user, 40, volume=50))
 				if(!can_dig(user))
 					return TRUE
-				to_chat(user, "<span class='notice'>You dig a hole.</span>")
+
+				if(W.tool_behaviour == TOOL_MINING && rand(1,15) == 1) //this may be too often - in testing 20 felt like too rare to be worth exploring
```
Comment (PKPenguin321) [inline code/game/turfs/simulated/floor/plating/asteroid.dm]: `prob()` is better used here than rand, I think prob(7) gives roughly the same odds as what you've put

Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: oh thanks, that sounds cleaner too. will fix in a bit

**Comment by zxaber** [comment]
Does gas stick around on Lavaland, or does evaporate and disappear like it does in space?

I suppose walling a large square on Lavaland with scrubbers leading to filters would be fun if there's anything worthwhile to find.

**Comment by read-0nly** [comment]
It dissipates pretty quickly - even a flaming tritium fissure's heat ends up pretty self-contained. basalt tiles have planetary atmos so the settle back to their initial gas mix

**Comment by zxaber** [comment]
So is there any way to really harvest it?

**Comment by read-0nly** [comment]
quench the fire, floor-tile and wall it in, use the white air filters to take what you want - or maybe just build a gas lab around the fissure

File: code/game/turfs/simulated/floor/plating/asteroid.dm (hunk 1 above)
Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: not sure I used prob right, testin g it now then gonna figure out how to correct it here lol

**Comment by zxaber** [comment]
I don't know why I was thinking it would just make an initial burst of gas and then stop. I get it now; spews gas constantly.

Good PR.

**Comment by read-0nly** [comment]
Thanks! I hope it'll be fun - I nerfed the gasses quite a bit but had some pretty happy accidents testing them at first

**Comment by read-0nly** [comment]
added unexpected bonus - if a fissure fire gets too hot and wild it'll just turn it to lava and stop the fire. So a tritium fissure shouldn't turn lavaland into the sun

File: code/game/turfs/simulated/floor/plating/asteroid.dm [hunk 2]
```diff
@@ -68,7 +68,27 @@
...
+						if(1,2)
+							initial_gas_mix = "plasma=32;n2=17;TEMP=360"
+						if(4,3,5)
+							initial_gas_mix = "co2=33;n2=7;TEMP=450"
+						if(6)
+							initial_gas_mix = "tritium=24;co2=20;no2=2;TEMP=360"
+						if(7)
+							initial_gas_mix = "nob=12;pluox=7;n2=32;water_vapor=70;TEMP=170"
+						if(8)
+							initial_gas_mix = "pluox=32;stim=13;TEMP=295"
```
Comment (Tlaltecuhtli) [inline code/game/turfs/simulated/floor/plating/asteroid.dm]: haha who needs fusion when you can dig while dragging a scrubber

Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: the gas mixes probably need work but I didn't know what I was doing so I played it by ear and tossed in what I thought might be fun to find. If you have suggestions on what the gas mixes should be I'm not committed to these

**Comment by anconfuzedrock** [comment]
do shovels get more or less fissures?

**Comment by read-0nly** [comment]
They get none, though honestly I have no better excuse than "you don't crack rocks with shovels silly" - the metaphor is cracking the bedrock under the basalt and breaking into trapped gas.

Arguably shoveling with a pickaxe is as ridiculous an idea though so there's no reason the shovel wouldn't do it too.

**Comment by Militaires** [comment]
this seems easily spammable and a bad idea, what would be a better idea is to sprinkle gas fissures across the map like tendril spawners and have GPS point to their locations.

**Comment by granpawalton** [comment]
Do you think you could make it so it spawns nitryl and bz instead of nob and stimulum? Nob can do a lot even with like only .5 of a mole and both in general kinda belong behind a time gate

If you really want those two to be obtainable outside of atmos/sci i think a tool/structure for gathering it that is produceable in cargo via credits is a good way of time gating it

**Comment by dr-bee** [comment]
Giving mining easy access to the best stun reduction item in the game is not a good idea. Love the idea otherwise, gives a reason for an engineer to head down to lavaland and build a gas mining plant.

**Comment by read-0nly** [comment]
I tried to implement a bit of everyone's ideas - however I'm not sure how I'd go about doing something like the tendrils. But since the idea is that this is supposed to be random and also not entirely a boon I reduced probability and played with the gas mixes a bit to make them less rich but also more interesting and messy, as a way to try to bypass the spammability of it. Is this better?

File: code/game/turfs/simulated/floor/plating/asteroid.dm [hunk 3]
```diff
@@ -68,7 +68,23 @@
...
 				if(!can_dig(user))
 					return TRUE
-				to_chat(user, "<span class='notice'>You dig a hole.</span>")
+
+				if(W.tool_behaviour == TOOL_MINING && prob(2)) //I forgot about better mining tools. Brought down the probability and making the options more likely to be boring
+					to_chat(user, "<span class='warning'>You opened a fissure! A gas is escaping...</span>")
+					icon_state = "fissure"
+					icon_plating = "fissure"
+					environment_type="fissure"
+					switch(rand(1,10)) //This probably needs to be adjusted - picks between different types of fissures
```
Comment (Militaires) [inline code/game/turfs/simulated/floor/plating/asteroid.dm]: also pickweight would be better than this

**Comment by HomocidlePancakes** [comment]
This PR actually adds some functionality to aux base besides "Unga dunga lets rebuild every machine from station on lavaland unga dunga". How about giving atmostechies access to aux base (if they don't already) to set up a gas mining station?

File: code/game/turfs/simulated/floor/plating/asteroid.dm (hunk 2 above)
Comment (Tlaltecuhtli) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: anything that isnt nb or stimulum is fine

Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: yeah i reworked them, made the mixes a bit less rich but more things mixed together and also always above ignition except for the cold pluoxy fissure to make them a bit more dangerous to look for

File: code/game/turfs/simulated/floor/plating/asteroid.dm (hunk 3 above)
Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: Oh hey pickweight is a thing. Testing it now, rand was admittedly pretty hacky

Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: Corrected in the last commit

**Comment by read-0nly** [comment]
I like that idea - At least science has access so they can go make bombs and things

**Comment by as334** [comment]
I think a Miasma one could be good, sorta like a swamp gas kinda deal.

File: code/game/turfs/simulated/floor/plating/asteroid.dm [hunk 4]
```diff
@@ -68,7 +68,15 @@
...
 				if(!can_dig(user))
 					return TRUE
-				to_chat(user, "<span class='notice'>You dig a hole.</span>")
+
+				if(W.tool_behaviour == TOOL_MINING && prob(2)) //I forgot about better mining tools. Brought down the probability and making the options more likely to be boring
+					to_chat(user, "<span class='warning'>You opened a fissure! A gas is escaping...</span>")
+					icon_state = "fissure"
+					icon_plating = "fissure"
+					environment_type="fissure"
+					initial_gas_mix = pickweight(fissure_gas_mix)
```
Comment (ShizCalev) [inline code/game/turfs/simulated/floor/plating/asteroid.dm]: `fissure_gas_mix` doesn't exist on this obj, this should be failing... hm.

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/game/turfs/simulated/floor/plating/asteroid.dm [hunk 5]
```diff
@@ -92,6 +100,10 @@
 	environment_type = "basalt"
 	floor_variance = 15
 	digResult = /obj/item/stack/ore/glass/basalt
+	/var/list/fissure_gas_mix = list("plasma=42;n2=32;TEMP=600" = 3,
```
Comment (ShizCalev) [inline code/game/turfs/simulated/floor/plating/asteroid.dm]: remove the first `/`

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/game/turfs/simulated/floor/plating/asteroid.dm (hunk 5 above)
Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: removed the slash, does it also fix the var not existing above?

Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: nope now it doesn't compile. welp, debugging and readings docs, fix shortly

Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: Well it's fixed but I really didn't want to modify the base asteroid tile. I hadn't realized the dig behavior was declared at the asteroid level. I think I need to approach this differently, I had a worst grasp of this than I thought. Not sure how to proceed, I should probably retract this while I hammer out a better solution. I guess I leave it to your collective wisdom - does this work, or should I start over and do it properly and put everything under basalt?

Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: actually - thinking about it again- finding gas vents on asteroids could also be pretty interesting.  I'm satisfied with this, Was there anything else you'd like me to review ShizCalev? It should work as expected now

File: code/game/turfs/simulated/floor/plating/asteroid.dm (hunk 4 above)
Comment (read-0nly) [reply code/game/turfs/simulated/floor/plating/asteroid.dm]: Fixed the declaration and moved it to the root of asteroid - more details in other request so I'll go ahead and try to resolve this one

**Comment by ShizCalev** [comment]
Just be mindful this will also affect snow turf (/turf/open/floor/plating/asteroid/snow), in places such as Snowdin & VR.

**Comment by tralezab** [comment]
You could have different turfs have different kinds of fissures, the snow one having super cold gases

**Comment by Tlaltecuhtli** [comment]
a vent that makes x gas overtime and you have to scrub it and pipe it back to mining base is more interestint

**Comment by tralezab** [comment]
imagine if there was a new job that had to hook up and move gases from lavaland to the atmos on the station

i'd play it

**Comment by read-0nly** [comment]
Thanks for the suggestions, everyone, including the Miasma one (I really like that idea, I keep forgetting Miasma exists lol - It'd fit in well with the idea of it being a sort of hot mess of mixed gasses)

Putting it on a delay would make some sense, though I'm not sure how one would get a middle-ground between always pumping like it is now and just dumping a bunch of gas in the current tile on a tick

Thanks for pointing out snow turf, I'm still very green and maybe a bit overeager lol I didn't realize it inherited from asteroid. Implementing different gas mix lists for different turf types and making the base asteroid gas mix list empty, then adding an empty check, like tralezab suggested sounds like a good way to fix it. I'm at work right now and have a pretty tight schedule during the week but will try to hammer out a new commit when I get home to implement this stuff.

**Comment by Cyprex** [comment]
To give you an idea of how to get the gases on the station: portable air scrubbers or pumps are by far the easiest way, though I'm not sure if they can endure hot fires. They also store only a limited amount, might be slow, and are only available in a limited amount. To use pipes+syphons you need to wall+floor in the area to create an area definition(for power), and then remove the floors to expose the actual basalt.

**Comment by ExcessiveUseOfCobblestone** [comment]
That list needs to have a more global presence so we're not making a list for literally every tile of asteroid.

**Comment by AnonymousTGplayer** [comment]
I disagree on inclusion of the pluoxium/stimulum fissure. Stimulum is generally considered an end-game gas, requiring quite some effort to produce to begin with. Including it as a random chance could lead to power creep. Another issue is that tritium is radioactive, meaning that if the fissure is left unattended, it would slowly cause Lavaland to become a radioactive wasteland.

**Comment by granpawalton** [comment]
@AnonymousTGplayer he already took that and the other problematic one out and the radiation issue wouldnt be useful to a traitor or something easy to create by accident

**Comment by Cyprex** [comment]
The radiation issue might cause performance issues, we'll see.

**Comment by read-0nly** [comment]
I've globalized the list and it's now a nested list that allows each environment type to have its own set of possible gas mixes

**Comment by read-0nly** [comment]
Uh, how do I fix this? It builds ok, so I think it's more about it not following code standards.
"Unmanaged global var use detected in code, please use the helpers."

File: code/game/turfs/simulated/floor/plating/asteroid.dm [hunk 6]
```diff
@@ -1,6 +1,18 @@
 
 /**********************Asteroid**************************/
 
+//Fissure mixes per tile class
+var/list/fissure_gas_mix = list(
```
Comment (AnturK) [inline code/game/turfs/simulated/floor/plating/asteroid.dm]: Use GLOBAL_LIST_INIT helper for this.

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by AnturK** [comment]
I'm not sure how i feel about this. I got nothing against the feature but i really don't like adding more atmos churn what with miasma already being a thing.

**Pull request closed without merging**